File: ui/src/store/reducers/guides.ts
// Twin of `faf_domain::state::guides::reduce`. Pinned by the conformance
// fixture, so a transition that drifts from the Rust one fails a test rather
// than shipping.

import type { GuidesEvent, GuidesState } from "../../ipc/bindings";

/** Twin of `settle`: drop a decided row and remember that it is decided. */
function settle(state: GuidesState, number: number): GuidesState {
  return {
    ...state,
    submissions: state.submissions.filter((held) => held.number !== number),
    settled: state.settled.includes(number) ? state.settled : [...state.settled, number],
  };
}

export function reduceGuides(state: GuidesState, event: GuidesEvent): GuidesState {
  switch (event.type) {
    case "configured": {
      const { repo, configured } = event.payload;
      // Only moves *into* unconfigured, and only from signed out: a restored
      // session must not be thrown away by a later configuration event.
      const auth =
        !configured && state.auth.type === "signedOut" ? ({ type: "unconfigured" } as const) : state.auth;
      return { ...state, repo, auth };
    }
    case "signInStarted":
      return { ...state, auth: { type: "waiting", payload: { login: event.payload.login } } };
    case "signedIn":
      return { ...state, auth: { type: "signedIn", payload: { identity: event.payload.identity } } };
    case "signInFailed":
      return { ...state, auth: { type: "failed", payload: { reason: event.payload.reason } } };
    case "signInCancelled":
    case "signedOut":
      return { ...state, auth: { type: "signedOut" } };
    case "queueLoading":
      return { ...state, status: { type: "loading" } };
    case "queueLoaded":
      return {
        ...state,
        // A submission this session already decided does not come back just
        // because GitHub's list has not caught up with the close.
        submissions: event.payload.submissions.filter(
          (submission) => !state.settled.includes(submission.number),
        ),
        status: { type: "ready" },
      };
    case "queueLoadFailed":
      return { ...state, status: { type: "failed", payload: { reason: event.payload.reason } } };
    case "accepting":
      return { ...state, write: { type: "accepting", payload: { number: event.payload.number } } };
    case "rejecting":
      return { ...state, write: { type: "rejecting", payload: { number: event.payload.number } } };
    // A settled verdict drops the row it settled. The queue is "what is still
    // open", and leaving a decided submission in it invites a second verdict.
    case "accepted":
      return {
        ...settle(state, event.payload.number),
        write: { type: "accepted", payload: { number: event.payload.number } },
      };
    case "rejected":
      return {
        ...settle(state, event.payload.number),
        write: { type: "rejected", payload: { number: event.payload.number } },
      };
    case "writeFailed":
      return {
        ...state,
        write: {
          type: "failed",
          payload: { number: event.payload.number, reason: event.payload.reason },
        },
      };
    // A fresh submission is being written, so the last one's result is no
    // longer the answer to anything on screen.
    case "submitReset":
      return { ...state, submit: { type: "idle" } };
    case "submitting":
      return { ...state, submit: { type: "sending" } };
    case "submitted":
      return { ...state, submit: { type: "sent", payload: { url: event.payload.url } } };
    case "submitFailed":
      return { ...state, submit: { type: "failed", payload: { reason: event.payload.reason } } };
  }
}

File: crates/faf-app/src/services/guides.rs
//! Catalogue maintenance: signing in to GitHub, and the submission queue.
//!
//! The interesting part is concurrency policy, because three of these
//! operations have very different shapes:
//!
//! - **Signing in polls for minutes.** It runs single-flight, so pressing the
//!   button twice cannot leave two loops polling the same code, and cancelling
//!   reaches the loop through the port rather than by dropping a task.
//! - **A verdict is a short write that must not overtake another.** Two accepts
//!   at once would both read the catalogue, both patch their own copy and one
//!   would lose; the sha check turns that into a refusal rather than a silent
//!   overwrite, but serialising them means it never happens in the first place.
//! - **Reading the queue is replaceable.** The newest answer wins; an older one
//!   still in flight is not worth waiting for.
//!
//! Every write re-reads the queue afterwards, for the reason the tournament
//! service does the same: the response says nothing about what else changed,
//! and a list that disagrees with the server is worse than a slow one.

use faf_domain::state::{GuidesCommand, GuidesEvent};

use crate::runtime::{EventSink, ServiceCtx};

pub async fn handle(cmd: GuidesCommand, ctx: &ServiceCtx, out: &EventSink) {
    match cmd {
        GuidesCommand::Restore => restore(ctx, out).await,
        GuidesCommand::SignIn => sign_in(ctx, out).await,
        GuidesCommand::CancelSignIn => {
            ctx.ports.guides.cancel_login();
            out.emit(GuidesEvent::SignInCancelled);
        }
        GuidesCommand::SignOut => {
            ctx.ports.guides.sign_out().await;
            out.emit(GuidesEvent::SignedOut);
        }
        GuidesCommand::LoadQueue => load_queue(ctx, out).await,
        GuidesCommand::Accept { number } => accept(number, ctx, out).await,
        GuidesCommand::Reject {
            number,
            reason,
            note,
        } => reject(number, reason, note, ctx, out).await,
        GuidesCommand::Submit { draft } => {
            out.emit(GuidesEvent::Submitting);
            // The author is this client's FAF account, which is what a reader
            // of the catalogue will see credited. GitHub knows who opened the
            // issue; the catalogue entry should name the player.
            let author = out.with_state(|state| {
                state
                    .auth
                    .player
                    .as_ref()
                    .map(|player| player.name.clone())
                    .unwrap_or_default()
            });
            let entry = faf_domain::state::entry_from_draft(&draft, &author);
            match ctx.ports.guides.submit(entry, draft.body.clone()).await {
                Ok(url) => {
                    out.emit(GuidesEvent::Submitted { url });
                    // The queue the author is about to look at should already
                    // contain what they just sent.
                    load_queue(ctx, out).await;
                }
                Err(reason) => out.emit(GuidesEvent::SubmitFailed { reason }),
            }
        }
    }
}

/// Announce what the client was configured with, and pick up a stored session.
///
/// Runs when the tab first opens. The configuration event goes out first and
/// unconditionally: whether signing in is possible at all is what the UI needs
/// before it can decide between a button and an explanation.
async fn restore(ctx: &ServiceCtx, out: &EventSink) {
    out.emit(GuidesEvent::Configured {
        repo: ctx.ports.guides.repo(),
        configured: ctx.ports.guides.configured(),
    });
    match ctx.ports.guides.restore_login().await {
        Ok(Some(identity)) => out.emit(GuidesEvent::SignedIn {
            identity: Box::new(identity),
        }),
        // Nobody has ever signed in here. Not a failure and not worth a word.
        Ok(None) => {}
        // There was a session and it no longer works. Said out loud, because
        // otherwise an expired token looks exactly like never having signed in.
        Err(reason) => out.emit(GuidesEvent::SignInFailed { reason }),
    }
}

async fn sign_in(ctx: &ServiceCtx, out: &EventSink) {
    let Some(_guard) = ctx.guides_login_active.try_acquire() else {
        // Already waiting on a code. Starting a second one would issue a
        // second code and leave the one on screen dead.
        return;
    };

    let code = match ctx.ports.guides.begin_login().await {
        Ok(code) => code,
        Err(reason) => {
            out.emit(GuidesEvent::SignInFailed { reason });
            return;
        }
    };

    out.emit(GuidesEvent::SignInStarted {
        login: Box::new(faf_domain::state::DeviceLogin {
            user_code: code.user_code.clone(),
            verification_uri: code.verification_uri.clone(),
            expires_at: super::now_seconds().saturating_add(code.expires_in),
        }),
    });

    match ctx.ports.guides.complete_login(code).await {
        Ok(identity) => {
            out.emit(GuidesEvent::SignedIn {
                identity: Box::new(identity),
            });
            // Now that there is a token the queue reads under a much larger
            // rate limit, and the maintainer is about to act on it.
            load_queue(ctx, out).await;
        }
        // A cancellation already emitted its own event; anything else is worth
        // reporting where the sign-in button is.
        Err(reason) if reason.contains("cancelled") => {}
        Err(reason) => out.emit(GuidesEvent::SignInFailed { reason }),
    }
}

async fn load_queue(ctx: &ServiceCtx, out: &EventSink) {
    let generation = ctx.guides_queue_generation.begin();
    out.emit(GuidesEvent::QueueLoading);
    let answer = ctx.ports.guides.list_submissions().await;
    if !ctx.guides_queue_generation.is_current(generation) {
        return; // A newer load has already been asked for.
    }
    match answer {
        Ok(submissions) => out.emit(GuidesEvent::QueueLoaded { submissions }),
        Err(reason) => out.emit(GuidesEvent::QueueLoadFailed { reason }),
    }
}

async fn accept(number: i32, ctx: &ServiceCtx, out: &EventSink) {
    let _order = ctx.guides_verdict.acquire().await;

    // Read back rather than carried on the command: the queue may have been
    // reloaded since the button was drawn, and publishing an entry that is no
    // longer what the issue says would be worse than refusing.
    let Some(submission) = out.with_state(|state| state.guides.submission(number).cloned()) else {
        out.emit(GuidesEvent::WriteFailed {
            number,
            reason: "that submission is no longer in the queue".into(),
        });
        return;
    };
    if !submission.is_acceptable() {
        out.emit(GuidesEvent::WriteFailed {
            number,
            reason: "this submission carries no catalogue entry to publish".into(),
        });
        return;
    }

    out.emit(GuidesEvent::Accepting { number });
    match ctx.ports.guides.accept(submission).await {
        Ok(()) => {
            out.emit(GuidesEvent::Accepted { number });
            load_queue(ctx, out).await;
            // And the library, because the maintainer's next question is
            // whether it worked. Leaving them to press refresh on another tab
            // to find out is how a working write looks broken.
            super::training::handle(faf_domain::state::TrainingCommand::Load, ctx, out).await;
        }
        Err(reason) => out.emit(GuidesEvent::WriteFailed { number, reason }),
    }
}

async fn reject(
    number: i32,
    reason: faf_domain::state::RejectReason,
    note: String,
    ctx: &ServiceCtx,
    out: &EventSink,
) {
    let _order = ctx.guides_verdict.acquire().await;

    out.emit(GuidesEvent::Rejecting { number });
    match ctx.ports.guides.reject(number, reason, note).await {
        Ok(()) => {
            out.emit(GuidesEvent::Rejected { number });
            load_queue(ctx, out).await;
        }
        Err(reason) => out.emit(GuidesEvent::WriteFailed { number, reason }),
    }
}
